[뒤로](https://github.com/papamoomin/Study#a-tour-of-c)  

# 목차
[1. 소개](#Chap1)  
[2. 분할 컴파일](#Chap2)  
[3. 모듈(C++20) XXX](#Chap3)  
[4. 네임스페이스](#Chap4)  
[5. 에러 처리](#Chap5)  
[6. 함수 인자와 반환 값](#Chap6)




<br><br><br>


<a name="Chap1"></a>
# 1. 소개

c++은 독립적으로 개발된 여러 부분으로 구성된다.  
예를 들면 함수, 사용자 정의 타입, 클래스 계층 구조, 템플릿 등을 들 수 있다.  

이들을 관리하는 데에 핵심은 이들 간의 상호작용을 명확히 정의하는 것이다.  
이를 위해서는 각 구성 요소의 인터페이스와 구현을 분리하는 것이 첫 단계로, 선언을 바탕으로 인터페이스를 표현할 수 있다.  

선언은 함수나 타입을 사용하는 데 필요한 모든 사항을 명시한다.  
그리고 핵심은 어딘가에 존재하는 몸체. 즉, 정의이다.

함수를 비롯한 구성 요소를 여러 번 선언하는 것은 가능하지만, 정의는 오직 한 번만 가능하다.  

<br><br>

<a name="Chap2"></a>
# 2. 분할 컴파일

c++은 분할 컴파일을 지원한다.  
사용자 코드에서 사용할 타입과 함수의 선언만 존재하면 컴파일이 가능하며, 타입과 함수의 정의가 분리된 파일에서 각기 따로 컴파일할 수 있다.  
이를 바탕으로 프로그램을 반독립적(semi-independent) 코드 조각의 집합으로 조직화 가능하다.  
이런 분리는 컴파일에 들이는 시간을 최소화하며 논리적으로 구분되는 프로그램의 분리를 강제해 에러의 소지도 감소한다.  
이렇게 분할 컴파일된 코드의 조각을 흔히 라이브러리라고 부른다.  

일반적으로 모듈의 인터페이스를 명시하는 선언을 파일에 담는다.  
그리고 그 파일의 이름은 모듈을 사용하고 있음을 의도적으로 드러낼 수 있도록 정한다.  

```cpp
// vector.h
class vector
{
public:                 vector(const int size);
public:                 ~vector();
public:     double&      operator[](const int index);
public:     int         size();

private:    double*     _element;
private:    int         _size;
};
```

```cpp
//vector.cpp
#include "vector.h"
#include <iostream>
using namespace std;

vector::vector(const int size) :_element(new double[size]), _size{size}
{
	for (int ii = 0; ii < size; ++ii)
	{
		cin >> _element[ii];
	}
}

vector::~vector()
{
	delete[] _element;
}

double& vector::operator[](const int index)
{
	return _element[index];
}

int vector::size()
{
	return _size;
}
```

```cpp
//main.cpp
#include "vector.h"
#include<iostream>
using namespace std;

double sumVector(vector& v)
{
	double sum = 0;
	for (int ii = 0; ii < v.size(); ++ii)
	{
		sum += v[ii];
	}
	return sum;
}

int main()
{
	vector v(5);
	cout << sumVector(v);
	return 0;
}
```

여기서 vector.h는 vector의 인터페이스이다.  
main.cpp는 #include"vector.h"로 vector를 사용한다.  
vector.cpp는 #include"vector.h"로 vector를 정의한다.  
main.cpp와 vector.cpp는 vector.h에 명시된 vector의 인터페이스를 공유하지만, 두 cpp는 따로 컴파일이 가능한 독립적인 파일이다.  

분할 컴파일은 언어 측면보다는 특정 구현의 장점을 잘 활용하는 실용적 측면에서 몹시 중요하다.  
프로그램을 조직화하는 가장 좋은 방법은 프로그램을 서로 간의 의존성이 잘 정의된 모듈의 집합으로 생각해야되며, 분할 컴파일이 가능하도록 파일 단위의 물리적 모듈화를 활용하여야 한다.  

이렇듯 독립적 컴파일이 가능한 cpp파일과 코드에서 #include하는 h파일을 각각 변환 단위(translation unit)이라고 부르며, 이 변환 단위들이 모여 한 프로그램을 구성한다.  

<br><br>

<a name="Chap3"></a>
# 3. 모듈 (c++20)
c++20에 관한 내용이니 당장 중요한 우선순위가 되지 못하기에 챕터만 만들어두고 넘어가겠다.

<br><br>

<a name="Chap4"></a>
# 4. 네임스페이스
c++에서는 네임스페이스라는 매커니즘을 이용해 일부 선언을 함께 묶고, 다른 이름과 충돌하지 않게 한다.  

```cpp
#include<iostream>
using namespace std;

namespace myCode
{
	int main();
}

int myCode::main()
{
	cout << "this is myCode's main" << endl;
	return 0;
}

int main()
{
	return myCode::main();
}
```
같은 이름에 같은 인수 타입과 갯수를 가진 함수더라도 다른 네임스페이스에 작성하여 기존의 네임스페이스와 충돌하지 않도록 보장해둘 수 있다.  
여기서 위의 main은 네임스페이스 myCode에 정의된 main이며, 기존에 늘 사용하던 아래의 main은 전역 네임스페이스에 정의된 main이다.  

다른 네임스페이스는 myCode::main()처럼 쉽게 접근할 수 있으며, 이름을 지정하는 일을 반복하기 싫을 경우 using 선언을 이용해 해당 이름을 현재 스코프에 가져올 수 있다.

```cpp
#include<iostream>

namespace myCode
{
	void say();
}

void myCode::say()
{
	//cout << "this is myCode's main" << endl;  //error
	using namespace std;
	cout << "this is myCode's main" << endl;
	return;
}

int main()
{
	//say();                                    //error
	//cout << "this is main" << endl;           //error
	using namespace myCode;
	say();
	return 0;
}
```
using 선언을 이용하면 대상 네임스페이스의 이름을 현재 스코프에 선언한 것처럼 사용할 수 있다.  
늘 상단에 사용하던 using namespace std; 도 네임스페이스 std를 사용하겠다는 의미.  

하지만 using을 남용하면 이름을 선택적으로 사용할 수가 없으므로 신중히 사용하도록 주의하자.  
using은 어플리케이션 전체에 사용하는 라이브러리가 있는 경우나, 네임스페이스를 쓰지 않는 어플리케이션을 개선할 때 쓰는 게 좋다.  
큰 규모의 프로그램 구성 요소를 조직화할 때는 네임스페이스를 사용하는 것이 좋으며, 이를 바탕으로 독립적으로 개발된 구성 요소로 이뤄진 프로그램을 단순화하여 만들 수 있다.  

<br><br>

<a name="Chap5"></a>
# 5. 에러 처리

에러 처리는 언어의 기능과 프로그래밍 기법, 도구를 아우르는 주제다.  
c++에서는 이를 돕는 몇 가지 기능을 제공한다.  

대부분의 c++ 구성 요소는 효율적인 추상화(ex. 사용자 정의 타입과 이를 활용한 알고리즘)를 설계하고 구현하기 위해 존재하며, 그 효과 중 하나가 런타임 에러를 탐지하는 지점과 처리하는 지점을 분리할 수 있다는 것이다.  

<br><br>

## 예외

2.3장 예제의 일부를 가져와 수정하겠다.
```cpp
double vector::sum(int startIndex, int lastIndex)
{
	double sum = 0;

	if ((startIndex < 0) || (_size <= lastIndex) || (lastIndex < startIndex))
	{
		throw out_of_range("out_of_range::vector::sum");
	}

	for (int ii = startIndex; ii <= lastIndex; ++ii)
	{
		sum += _element[ii];
	}

	return sum;
}

int main()
{
	vector v(5);
	try
	{
		cout << v.sum(0, 5);
	}
	catch (out_of_range& err)
	{
		cerr << err.what() << endl;
	}
	return 0;
}
```
throw는 vector::operator[]를 호출하는 임의의 함수에 존재하고 있는 out_of_range 타입의 예외 핸들러에게 제어권을 넘기는 것이다.  
이를 위해 해당 예외에 관심을 표하는 핸들러에 닿을 때까지 함수 호출 스택을 거슬러가며 필요한 만큼의 스코프와 함수를 종료시키고, 그 과정에서 소멸자 또한 호출시킨다.  

예외 처리가 필요한 코드는 try 블록에 작성하였다.  
그리고 try 절 내부에서 예외가 발생할 경우, 예외 핸들러를 제공하는 catch절에 진입한다.  
out_of_range 타입은 표준 라이브러리에 정의된다.  

catch 절의 인수로는 복사를 피하기 위해 참조를 사용하였다.  
throw에 설정한 메시지를 불러오기 위해 what 함수 또한 사용하였다.  

throw-try-catch는 간단하고 체계적이며 가독성이 높은 에러 처리를 가능하게 한다.  
하지만 과용은 금물. 이는 4장에서 설명할 예정이다.  

절대로 예외를 던지지 않는 함수는 noexcept로 선언한다.  
noexcept가 선언된 함수에서 예외를 던지면 std::terminate()가 호출되어 프로그램이 즉시 종료된다.  

<br><br>

## 불변 조건

위처럼 스코프 밖 접근에 대한 예외는 기본적 가정, 즉 사전 조건(precondition)이 성립하지 않으면 작동을 거부하는 방식의 예라 할 수 있다.  
함수를 정의할 때는 사전 조건이 무엇이며, 그 조건을 체크해야하는 지를 고려해야 한다.  

대부분의 사례에서는 간단한 불변 조건을 확인하는 것이 도움이 된다.  
위 예제에서도 operator[] 함수는 vector 타입의 객체에 작동하며, 적절한 값을 가지고 있지 않으면 무의미하다.  
특히 _element는 _size개의 double 배열이라는 점은 주석에나 명시할 수 있다.  
이처럼 클래스 수준에서 보장되어야 할 조건을 클래스 불변 조건(class invariant), 또는 불변 조건이라 부른다.  

클래스에 대한 불변 조건은 생성자가 보장하여야 하며, 멤버 함수는 스스로가 종료될 떄 불변 조건이 성립하는지 확인해야 한다.  
다시 2.3장 예제에서 예외처리를 추가해보자.  

```cpp
vector::vector(int size) :_element{ new double[size] }, _size{ size }
{
	if (size < 0)
	{
		throw length_error{ "length_error::vector::vector : 크기가 음수" };
	}

	for (int ii = 0; ii < _size; ++ii)
	{
		cin >> _element[ii];
	}
}

int main()
{
	try
	{
		vector v(-1);
	}
	catch (length_error& err)   // 크기가 음수인 경우
	{
		cerr << err.what() << endl;
        //throw;                // 예외를 다시 던질 수도 있음.
	}
	catch (bad_alloc & err)     // 메모리가 고갈된 경우
	{
		terminate();            // 프로그램 종료
	}
    return 0;
}
```
예외가 발생한 후 함수의 동작을 마무리할 수 없는 경우에는 최소한의 지역적 정리를 한 후 예외를 다시 던질 수도 있다.  

잘 설계된 코드일수록 try을 드물게 쓰고 RAII 기법을 체계적으로 사용한다. 이는 4장과 5장에서 나올 것이다.  

불변 조건이라는 것은 클래스 설계의 핵심적 역할을 하며, 함수에서는 사전 조건이 그와 비슷한 역할을 한다.  
불변 조건은 우리로 하여금 스스로 뭘 원하는지를 정확히 이해할 수 있게 도와주며 좀 더 구체적이 되도록 강제해 올바른 코드를 유지하는 데 도움을 준다.  
불변 조건의 개념은 생성자와 소멸자로 대표되는 c++의 자원 관리 매커니즘의 저변을 이룬다.  

## 여러 가지 에러 처리 방식

### 함수가 정해진 작업을 수행하지 못함을 나타내는 방법
- 예외를 던진다
- 실패를 나타내는 값을 반환한다
- 프로그램을 종료한다 (terminate, exit, abort 등의 함수 호출)

### 에러 지시자(코드)를 반환하는 경우
- 실패가 흔히 일어나고 예상이 가능한 경우 (ex. 파일 열기 요청)
- 직접적인 호출자가 실패를 처리할 수 있다는 기대가 합리적인 경우

### 예외를 던지는 경우
- 예외가 매우 드물게 발생해서 프로그래머가 에러 확인을 잊을 가능성이 큰 경우 (printf 등)
- 직접적인 호출자가 에러를 처리하기 어려운 경우. 이런 경우는 상위 호출자에게 에러를 전달한다 (어플리케이션의 모든 함수에서 네트워크 실패를 처리하리라 기대하긴 힘들다.)
- 하위 모듈에 새로운 종류의 에러가 추가될 가능성이 있어 상위 모듈이 새로운 에러를 처리하게 재작성하기 힘든 경우 (어플리케이션을 멀티 쓰레드로 바꾸거나 네트워크로 원격 자원에 접근하는 경우)
- 에러를 반환할 적당한 경로가 없는 경우. (ex. 생성자)
- 값 자체와 에러 지시자를 모두 반환해야 하는 경우(pair 등)에는 반환 경로가 복잡해지거나 비용이 커진다. 이런 경우 예외를 사용하지 않으면 출력 파라미터, 비지역적 에러 상태, 그 외의 임시 방편으로 이어진다.  
- 호출 체인을 통해 에러를 최상위 호출자에게 전달해야하는 경우, 비용이 크고 또 다른 에러의 소지도 높다.  
- 에러로부터의 복구가 여러 함수 호출 결과에 의존하는 경우. 예외를 사용하지 않으면 호출 사이사이에 지역적 상태를 유지해야 하고, 제어 구조가 복잡해진다.
- 에러를 탐지한 함수가 콜백인 경우. 어떤 함수가 호출되었는지조차 알기 어렵다.  
- 에러에 대해 '작업 취소'가 필요한 경우

### 프로그램을 종료하는 경우
- 복구할 수 없는 경우. (ex. 메모리 고갈)
- 심각한 에러가 발생해 스레드, 프로세스, 컴퓨터를 재시작하는 방식으로 에러를 처리하는 시스템인 경우.

프로그램의 종료를 보장하는 방법 중 하나는, 앞에서도 언급한 noexcept이다.  
이를 지정한 함수는 throw를 실행하면 terminate 함수를 호출한다.  

<br><br>

종료 조건은 논리적으로 명확하거나 쉽게 적용하기가 어렵다.  
미심쩍은 경우에는 예외를 사용하라.  

예외 처리가 느리다는 말이 있는데, 복잡하거나 드물게 발생하는 에러 조건을 처리하거나, 에러 코드를 반복해서 확인하는 것보다는 빠른 경우가 많다.  

예외를 이용해 간단하고 효과적인 에러 처리를 위해서는 4, 5장에 나올 RAII가 필수적이다.  
try를 남용한 코드는 에러 처리 전략의 나쁜 측면을 극명하게 보여준다.  

<br><br>

## 계약(C++20)
역시 C++20은 현재로서는 넘어가겠다.  

<br><br>

## 정적 어써션
예외는 실행 시간에 발생하는 에러를 보고하는 용도로 사용한다.  
하지만 컴파일 시간에 찾을 수 있는 에러라면 정적 어써션이 낫다.  
다음과 같이 정적 어써션을 이용해 컴파일 시간에 알 수 있는 속성을 확인하고, 예상에 벗어나면 컴파일러 에러의 형태로 문제를 보고할 수도 있다.  

```cpp
static_assert(4 <= sizeof(int), "친구들 만나느라 444"); // true
static_assert(5 <= sizeof(int), "친구들 만나느라 555"); // false, error
```
이 코드는 첫 인자의 값이 false인 경우, 컴파일 에러가 나며 두번째 인자의 메세지를 출력한다. 이처럼 기대 조건을 나타내는 구문을 어써션(assertion)이라고 한다.  

일반적으로 이렇게 static_assert(A,S) 구문은 A가 false면 S를 컴파일러 에러 메시지로 출력하며, S를 생략하면 컴파일러가 기본 메시지를 출력한다.  
기본 메시지는 일반적으로 static_assert가 호출된 소스코드의 위치와 주어진 조건을 문자열로 출력한다.  
제네릭 프로그래밍의 파라미터로 쓰이는 타입에 대한 어써션은 static_assert가 유용하다.  


<br><br>

<a name="Chap6"></a>
# 6. 함수 인자와 반환 값

